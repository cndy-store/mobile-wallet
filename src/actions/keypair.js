import { getItem, setItem, removeItem } from '../lib/keypairStorage';
import { StellarSdk } from '../lib/stellar';
const { Keypair, StrKey } = StellarSdk;

const STORAGE_KEY = 'STELLAR_SECRET';

export const KEYPAIR_STARTS_PROCESSING = 'KEYPAIR_STARTS_PROCESSING';
export const KEYPAIR_ERROR = 'KEYPAIR_ERROR';
export const KEYPAIR_LOAD = 'KEYPAIR_LOAD';
export const KEYPAIR_SAVE = 'KEYPAIR_SAVE';
export const KEYPAIR_REMOVE = 'KEYPAIR_REMOVE';

const chomp = str => str.replace(/\n|\r| /g, '');
const isValidSecret = secret => StrKey.isValidEd25519SecretSeed(secret);

export const keypairProcessing = bool => ({
  type: KEYPAIR_STARTS_PROCESSING,
  isProcessing: bool
});

export const keypairErrored = ({ error }) => ({
  type: KEYPAIR_ERROR,
  error
});

export const keypairLoadSucceded = ({ secret }) => {
  const keypair = secret ? Keypair.fromSecret(secret) : null;

  return {
    type: KEYPAIR_LOAD,
    keypair: keypair,
    error: null
  };
};

export const keypairSaveSucceded = ({ secret }) => {
  const keypair = Keypair.fromSecret(secret);

  return {
    type: KEYPAIR_SAVE,
    keypair: keypair,
    error: null
  };
};

export const keypairRemoveSucceded = () => ({
  type: KEYPAIR_REMOVE,
  keypair: null,
  error: null
});

export const loadKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));
    try {
      const secret = await getItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairLoadSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const saveKeypair = secret => {
  secret = chomp(secret);

  return async dispatch => {
    dispatch(keypairProcessing(true));

    if (!isValidSecret(secret)) {
      const error = new Error('Invalid Secret');
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }

    try {
      await setItem(STORAGE_KEY, secret);
      dispatch(keypairProcessing(false));
      dispatch(keypairSaveSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const removeKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));

    try {
      await removeItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairRemoveSucceded());
      return Promise.resolve();
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};
