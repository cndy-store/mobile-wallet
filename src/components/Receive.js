import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import {
  Icon,
  Container,
  Header,
  Left,
  Right,
  Body,
  Title,
  Content,
  Tab,
  Card,
  CardItem,
  Tabs
} from 'native-base';
import { Button, StyleSheet, View, Text } from 'react-native';
import QRCode from 'react-native-qrcode-svg';
import { encodePublicKey } from '../lib/keypairHelpers';

export class ReceiveScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {
      qrCodeWidth: null
    };

    this.setQRCodeWidth = this.setQRCodeWidth.bind(this);
  }

  setQRCodeWidth(event) {
    var { x, y, width, height } = event.nativeEvent.layout;

    this.setState({ qrCodeWidth: width });
  }

  renderQRCode() {
    if (!this.state.qrCodeWidth) return;

    const url = encodePublicKey(this.props.keypair.publicKey());
    return <QRCode size={this.state.qrCodeWidth} value={url} elc="M" />;
  }

  render() {
    return (
      <Content padder>
        <Card>
          <CardItem cardBody onLayout={this.setQRCodeWidth}>
            {this.renderQRCode()}
          </CardItem>
        </Card>
        <Card>
          <CardItem header>
            <Text>This is your public key in form of a QR Code</Text>
          </CardItem>
        </Card>
      </Content>
    );
  }
}

ReceiveScreen.propTypes = {};

const mapStateToProps = state => {
  return {
    keypair: state.keypair.keypair
  };
};

const mapDispatchToProps = dispatch => {
  return {};
};

export default connect(mapStateToProps, mapDispatchToProps)(ReceiveScreen);
