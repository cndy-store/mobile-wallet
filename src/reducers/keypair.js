import {
  KEYPAIR_STARTS_PROCESSING,
  KEYPAIR_ERROR,
  KEYPAIR_LOAD,
  KEYPAIR_SAVE,
  KEYPAIR_REMOVE
} from '../actions/keypair';

const defaultState = {
  isProcessing: false,
  error: null,
  keypair: null
};

export default function keypair(state = defaultState, action) {
  switch (action.type) {
    case KEYPAIR_STARTS_PROCESSING:
      return { ...state, isProcessing: action.isProcessing };
    case KEYPAIR_ERROR:
      return {
        ...state,
        error: action.error,
        isProcessing: action.isProcessing
      };
    case KEYPAIR_LOAD:
    case KEYPAIR_SAVE:
    case KEYPAIR_REMOVE:
      return {
        ...state,
        keypair: action.keypair,
        error: action.error,
        isProcessing: action.isProcessing
      };
    default:
      return state;
  }
}
