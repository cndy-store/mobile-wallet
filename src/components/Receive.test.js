import React from 'react';
import QRCode from 'react-native-qrcode-svg';
import { render } from '../__tests__/renderer';
import { Receive } from './Receive';
import { keypair } from '../__tests__/fixtures/keypair';

it('renders correctly', () => {
  const { toJSON } = render(<Receive keypair={keypair} />);
  expect(toJSON()).toMatchSnapshot();
});
