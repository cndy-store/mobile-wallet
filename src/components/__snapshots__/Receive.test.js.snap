// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  contentContainerStyle={
    Array [
      Object {
        "padding": 10,
      },
      undefined,
    ]
  }
  contentInset={
    Object {
      "bottom": 0,
    }
  }
  enableAutomaticScroll={true}
  enableResetScrollToCoords={true}
  extraHeight={75}
  extraScrollHeight={0}
  getScrollResponder={[Function]}
  handleOnScroll={[Function]}
  keyboardDismissMode="interactive"
  keyboardOpeningTime={250}
  keyboardShouldPersistTaps="handled"
  keyboardSpace={0}
  onScroll={[Function]}
  padder={true}
  resetKeyboardSpace={[Function]}
  resetScrollToCoords={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  scrollEventThrottle={1}
  scrollForExtraHeightOnAndroid={[Function]}
  scrollToEnd={[Function]}
  scrollToFocusedInput={[Function]}
  scrollToPosition={[Function]}
  showsVerticalScrollIndicator={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
  viewIsInsideTabBar={false}
  virtual={undefined}
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "borderBottomWidth": 0.5,
          "borderColor": "#ccc",
          "borderLeftWidth": 0.5,
          "borderRadius": 2,
          "borderRightWidth": 0.5,
          "borderTopWidth": 0.5,
          "elevation": 3,
          "flex": 1,
          "flexWrap": "nowrap",
          "marginBottom": 5,
          "marginLeft": 2,
          "marginRight": 2,
          "marginTop": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 1.5,
        }
      }
      virtual={undefined}
    >
      <View
        cardBody={true}
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 2,
            "flexDirection": "row",
            "paddingBottom": -5,
            "paddingLeft": -5,
            "paddingRight": -5,
            "paddingTop": -5,
          }
        }
        virtual={undefined}
      />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "borderBottomWidth": 0.5,
          "borderColor": "#ccc",
          "borderLeftWidth": 0.5,
          "borderRadius": 2,
          "borderRightWidth": 0.5,
          "borderTopWidth": 0.5,
          "elevation": 3,
          "flex": 1,
          "flexWrap": "nowrap",
          "marginBottom": 5,
          "marginLeft": 2,
          "marginRight": 2,
          "marginTop": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 1.5,
        }
      }
      virtual={undefined}
    >
      <View
        header={true}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderBottomWidth": null,
            "borderRadius": 2,
            "flexDirection": "row",
            "paddingBottom": 15,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 15,
          }
        }
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          This is your public key: 
          GCHSUHOKNZQAKTULV5YTMG5CKU2NKE7UUQVN6JT2JY7337CRBIAZ6LXK
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
